### PR TITLE
Catch exceptions raised while copying __dict__/__slots__ in BasePersi…

### LINKED ISSUE #2563: [BUG] 13.1+ persistence migration script raises an exception when converting UUID objects
### Steps to reproduce
1. run a bot using python-telegram-bot <13 (in my case, 12.8) and store some `uuid.UUID` instances in an user's `user_data`, and configure the bot to serialize temp data to a pickle file using `PicklePersistence`

2. use the [`PicklePersistence` migration script](https://gist.github.com/Bibo-Joshi/23b8c02249264256ded07661d0cd8a4f) linked in the wiki against your pickle file generated by ptb <13

### Expected behaviour
The pickle file migration is successfull

### Actual behaviour
Running the script raises an exception. Here's the full traceback:

```
Traceback (most recent call last):
  File "C:/Users/redacted/PycharmProjects/sticker-thief/convert_pp_13.1.py", line 92, in <module>
    persistence.get_user_data()
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 136, in get_user_data_insert_bot
    return instance.insert_bot(get_user_data())
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 310, in insert_bot
    return self._insert_bot(obj, {})
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 364, in _insert_bot
    new_obj[self._insert_bot(k, memo)] = self._insert_bot(val, memo)
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 364, in _insert_bot
    new_obj[self._insert_bot(k, memo)] = self._insert_bot(val, memo)
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 364, in _insert_bot
    new_obj[self._insert_bot(k, memo)] = self._insert_bot(val, memo)
  File "C:\Users\redacted\PycharmProjects\sticker-thief\venv\lib\site-packages\telegram\ext\basepersistence.py", line 369, in _insert_bot
    setattr(new_obj, attr_name, self._insert_bot(attr, memo))
  File "C:\Users\redacted\AppData\Local\Programs\Python\Python36-32\lib\uuid.py", line 224, in __setattr__
    raise TypeError('UUID objects are immutable')
TypeError: UUID objects are immutable
```

The exception is raised because UUID object don't want the user to use `__setattr__` (the exception is raised by `uuid.UUID().__setattr__()`), which is what the migration script explicitly uses at line 75 - but the exception is actually raised first, when the script uses `PicklePersistence().get_user_data()` (which calls `BasePersistence.insert_bot()`, which will try to recursively inject a new attribute to every object that has a `__dict__` attribute ([link to code](https://github.com/python-telegram-bot/python-telegram-bot/blob/52ce03929b3255eade1081b2ee6fa278708dbef9/telegram/ext/basepersistence.py#L382)), such as UUID instances)

Related thread in the group chat: https://t.me/pythontelegrambotgroup/487577

### Configuration
**Operating System:** Widnows 10

**Version of Python, python-telegram-bot & dependencies:**

```
$ python -m telegram
python-telegram-bot 13.6
Bot API 5.2
certifi 2019.09.11
Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 17:26:49) [MSC v.1900 32 bit (Intel)]
```

### Logs
Insert logs here (if necessary)

### PR DESCRIPTION
**Pull request by zeroone2numeral2**
…stence.replace/insert_bot()

Also updated the docstrings to reflect the changes in behavior with unexpected errors

closes #2563

### Checklist for PRs

- [ ] Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [x] Created new or adapted existing unit tests
- [x] Added myself alphabetically to `AUTHORS.rst` (optional)

### CODE REVIEW

**Comment by zeroone2numeral2** [comment]
I was looking at the contribution guidelines. Quiting:

> Add .. versionadded:: version, .. versionchanged:: version or .. deprecated:: version to the associated documentation of your changes, depending on what kind of change you made. This only applies if the change you made is visible to an end user. The directives should be added to class/method descriptions if their general behaviour changed and to the description of all arguments & attributes that changed.

I did made some changes to the docstring of the `replace_bot` and `insert_bot` methods, so I built the docs and looked for their docs source. But I can't find any `versionchanged` string. I'm not sure if I'm supposed to do something or not

**Comment by zeroone2numeral2** [comment]
Question about tests: I've added a `uuid.UUID` object  to `CustomClass`, which is immutable and would raise a `TypeError` in `insert/replace_bot`. I tried to run the tests as described in the contributions guidelines with `pytest -v`, but I'm a bit lost: the test for `test_persistence.py` failed with the following output: https://nekobin.com/loqazeruxi

I'm not entirely sure how I should update the test

File: telegram/ext/basepersistence.py [hunk 1]
```diff
@@ -278,20 +279,30 @@ def _replace_bot(cls, obj: object, memo: Dict[int, object]) -> object:  # pylint
...
+                return new_obj
+            if hasattr(obj, '__dict__'):
+                for attr_name, attr in new_obj.__dict__.items():
+                    setattr(new_obj, attr_name, cls._replace_bot(attr, memo))
+                memo[obj_id] = new_obj
+                return new_obj
+        except Exception as e:
+            exception_description = str(e)
+            warnings.warn(
+                f'Parsing of an object failed with the following exception: {exception_description}. '
```
Comment (Bibo-Joshi) [inline telegram/ext/basepersistence.py:300]: just `f'... {e}'` should work, too, shouldn't it? same below

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the PR!

Regarding docstrings: Those look fine, shouldn't be a problem to build them 👌 
Regarding tests: I forgot that pytest treats warnings as errors if not handled otherwise … On a  closer look I see that we have a for unpickable objects test [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/52ce03929b3255eade1081b2ee6fa278708dbef9/tests/test_persistence.py#L643-L690). You can copy that and replace `lock` with `uuid` (you can skip the `CustomClass` part and revert your current changes to the tests)

Also, `pre-commit` complains about a few things. Please make sure to run `pre-commit install` on your machine so that `pre-commit` can run on every commit. You can also have it run manually via `pre-commit run`. Have a look at the contribution guide for more details :)

File: telegram/ext/basepersistence.py [hunk 2]
```diff
@@ -278,20 +279,31 @@ def _replace_bot(cls, obj: object, memo: Dict[int, object]) -> object:  # pylint
...
+                return new_obj
+            if hasattr(obj, '__dict__'):
+                for attr_name, attr in new_obj.__dict__.items():
+                    setattr(new_obj, attr_name, cls._replace_bot(attr, memo))
+                memo[obj_id] = new_obj
+                return new_obj
+        except Exception as exception:
+            warnings.warn(
+                f'Parsing of an object failed with the following exception: {exception}. '
+                f'See the docs of BasePersistence.insert_bot for more information.',
```
Comment (harshil21) [inline telegram/ext/basepersistence.py:302]:
```suggestion
                f'See the docs of BasePersistence.replace_bot for more information.',
```

**Comment by Bibo-Joshi** [comment]
Except for the typo, this looks good to me :)

**Comment by Bibo-Joshi** [comment]
Thanks for the contribution @zeroone2numeral2 :)

**Comment by zeroone2numeral2** [comment]
Thanks @harshil21 for the commit! I've been busy during the weekend

And thanks for the merge :)

**Pull request merged**